deep_qa/data/instances/mc_tuples_with_background_instance.py

```diff
@@ -0,0 +1,358 @@
+        tuple 2 or more.
+
+        If a tuple slot is reserved for context, it contains "context:" at the start.
+
+        Some tuples contain coreference resolution info (from one of the table sources).  We do not
+        currently handle this, but plan to eventually replace the referring expression with its resolved
+        value.
+            Example: it[sky]|is|blue --> would become --> sky|is|blue
+
+        Some tuples additionally contain prefixed information about some of the "objects" such as "L:" for
+        locations and T: for time.  At present this information is simply filtered out, but can be used
+        later.
+        """
+        fields = line.split("\t")
+        assert len(fields) == 5, \
+            "Input line is not the correct number of fields (should be 5, it is {0}) " \
+            "-- line:{1}".format(len(fields), line)
+        # Question index
+        index = int(fields[0])
+        # Number of answer candidates
+        num_answers = int(fields[1])
+        # Answer Candidate tuples
+        answers = fields[2].split("###")
+        answer_tuples = []
+        for curr_answer in answers:
+            curr_answer_tuple_strings = curr_answer.split("$$$")
+            curr_answer_tuples = [TextTuple.from_string(tuple_string)
+                                  for tuple_string in curr_answer_tuple_strings if len(tuple_string) > 0]
+            answer_tuples.append(curr_answer_tuples)
+        assert len(answer_tuples) == num_answers, \
+            "ERROR: {0} answers found, {1} specified.".format(len(answer_tuples), num_answers)
+        # Background tuples
+        background_tuple_strings = fields[3].split("$$$")
+        background_tuples = [TextTuple.from_string(background_string)
+                             for background_string in background_tuple_strings if len(background_string) > 0]
+        # Label
+        label = int(fields[4])
+        assert label < num_answers
+
+        return cls(answer_tuples, background_tuples, label=label, index=index)
+
+
+class IndexedMCTuplesWithBackgroundInstance(IndexedInstance):
+
+    def __init__(self,
+                 answers_indexed: List[List[List[List[int]]]],
+                 background_indexed: List[List[List[int]]],
+                 label,
+                 index: int=None):
+        super(IndexedMCTuplesWithBackgroundInstance, self).__init__(label, index)
+        self.answers_indexed = answers_indexed
+        self.background_indexed = background_indexed
+
+    @classmethod
+    @overrides
+    def empty_instance(cls):
+        return IndexedMCTuplesWithBackgroundInstance([], [], label=None, index=None)
+
+    @overrides
+    def get_lengths(self) -> Dict[str, int]:
+        # We care only about the longest slot here because all slots will be padded to the same length.
+        max_answer_length = 0
+        max_num_slots = 0
+        max_question_tuples = 0
+        for curr_answer in self.answers_indexed:
```

Comment:
I'm not a huge fan of prepending `curr_` to variables like this - you can just use pluralization most of the time to get a different (and easy to understand) variable name, and `curr` is a shortening of a word, which you should almost always avoid in variable names.  I would use `answer` here, and `answer_slots` and `answer_length` for the variables below.